test_common/harness/kernelHelpers.cpp

```diff
@@ -1694,6 +1701,53 @@ Version get_max_OpenCL_C_for_context(cl_context context)
     return current_version;
 }
 
+bool device_supports_cl_c_version(cl_device_id device, Version version)
+{
+    auto device_cl_version = get_device_cl_version(device);
+
+    // In general, a device does not support an OpenCL C version if it is <=
+    // CL_DEVICE_OPENCL_C_VERSION AND it does not appear in the
+    // CL_DEVICE_OPENCL_C_ALL_VERSIONS query.
+
+    // If the device version >= 3.0 it must support the
+    // CL_DEVICE_OPENCL_C_ALL_VERSIONS query, and the version of OpenCL C being
+    // used must appear in the query result if it's <=
+    // CL_DEVICE_OPENCL_C_VERSION.
+    bool version_in_all_versions_query = false;
+    if (device_cl_version >= Version{ 3, 0 })
+    {
+        size_t opencl_c_all_versions_size_in_bytes{};
+        auto error =
+            clGetDeviceInfo(device, CL_DEVICE_OPENCL_C_ALL_VERSIONS, 0, nullptr,
+                            &opencl_c_all_versions_size_in_bytes);
+        test_error_ret(
+            error, "clGetDeviceInfo failed for CL_DEVICE_OPENCL_C_ALL_VERSIONS",
+            (false));
+        std::vector<cl_name_version> name_versions(
+            opencl_c_all_versions_size_in_bytes / sizeof(cl_name_version));
+        error = clGetDeviceInfo(device, CL_DEVICE_OPENCL_C_ALL_VERSIONS,
+                                opencl_c_all_versions_size_in_bytes,
+                                name_versions.data(), nullptr);
+        test_error_ret(
+            error, "clGetDeviceInfo failed for CL_DEVICE_OPENCL_C_ALL_VERSIONS",
+            (false));
+
+        for (const auto &name_version : name_versions)
+        {
+            Version current_version{ CL_VERSION_MAJOR(name_version.version),
+                                     CL_VERSION_MINOR(name_version.version) };
+            if (current_version == version)
+            {
+                version_in_all_versions_query = true;
+                break;
+            }
+        }
+    }
+
+    return version >= get_device_cl_c_version(device)
+        || version_in_all_versions_query;
```

Comment:
It may just be a bit too late in the day, but is the first condition backwards?  Seems like the device would support the requested version if `version <= get_device_cl_c_version(device)` instead?

Also, it's fairly minor, but I'd consider swapping these two boolean conditions also, since there's no need to query `CL_DEVICE_OPENCL_C_VERSION` if the version was already found in `CL_DEVICE_OPENCL_C_ALL_VERSIONS`.

Reply:
Yes, you're totally right. Good catch.

Agreed regarding the second change. Rather than store the result of checking CL_DEVICE_OPENCL_C_ALL_VERSIONS, I have changed it to early exit in the for loop instead.